unlock_indicator.c

```diff
@@ -91,6 +94,99 @@ static double scaling_factor(void) {
     return (dpi / 96.0);
 }
 
+/*
+ * Returns the cairo_font_face_t for sans-serif
+ * if this thing returns NULL, then we're somehow on a system without a
+ * sans-serif font, or a system without fontconfig (but with cairo somehow)
+ * Reference counts and caches the font-face, so we don't have memory leaks
+ * (and so that we don't have to go through all the parsing each time)
+ */
+
+static cairo_font_face_t *get_font_face() {
+    if (sans_serif) {
+        return sans_serif;
+    }
+
+    FcResult result;
+    /*
+     * Loads the default config.
+     * On successive calls, does no work and just returns true.
+     */
+    if (!FcInit()) {
+        DEBUG("Fontconfig init failed. No text will be shown.\n");
+        return NULL;
+    }
+    /*
+     * converts a font face name to a pattern for that face name
+     */
+    FcPattern *pattern = FcNameParse((const unsigned char *)"sans-serif");
+    if (!pattern) {
+        DEBUG("no sans-serif font available\n");
+        return NULL;
+    }
+
+    /*
+     * Gets the default font for our pattern. (Gets the default sans-serif font face)
+     * Without these two calls, the FcFontMatch call will fail due to FcConfigGetCurrent()
+     * not giving it a valid/useful config.
+     */
+    FcDefaultSubstitute(pattern);
+    if (!FcConfigSubstitute(FcConfigGetCurrent(), pattern, FcMatchPattern)) {
+        DEBUG("config sub failed?\n");
+        return NULL;
+    }
+    /*
+     * Looks up the font pattern and does some internal RenderPrepare work,
+     * then returns the resulting pattern that's ready for rendering.
+     */
+    FcPattern *pattern_ready = FcFontMatch(FcConfigGetCurrent(), pattern, &result);
+    FcPatternDestroy(pattern);
+    pattern = NULL;
+    if (!pattern_ready) {
+        DEBUG("no sans-serif font available\n");
+        return NULL;
+    }
+
+    /*
+     * Passes the given pattern into cairo, which loads it into a cairo freetype font face.
+     * Increment its reference count and cache it.
+     */
+    cairo_font_face_t *face = cairo_ft_font_face_create_for_pattern(pattern_ready);
+    FcPatternDestroy(pattern_ready);
+    sans_serif = cairo_font_face_reference(face);
+    return face;
+}
```

Comment:
I do not really know "font stuff" well, but why not replace the above function with
```C
if (sans_serif == NULL)
    sans_serif = cairo_toy_font_face_create("sans-serif",  CAIRO_FONT_SLANT_NORMAL, CAIRO_FONT_WEIGHT_NORMAL);
return sans_serif;
```
If "memory leaks" is the reason, then I would rather fix that instead of working around it. I am pretty sure that using the toy font backend should not cause memory leaks. And the reason that the toy font API is marked as toy are AFAIK things like right-to-left text and complicated scripts, not memory leaks.

Reply:
Doing that'll cause some memory leaks in `fontconfig`. The only way to avoid them is to make the `cairo_debug_reset_static_data` call, which can't be used in conjunction with our own caching.

It definitely seems like it's better to just let i3lock handle the fairly simple lookup + caching than to have to flush the static data every draw tick.

Reply:
These are the specific memory leaks I get from ASAN:

```=================================================================
==15286==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 256 byte(s) in 1 object(s) allocated from:
    #0 0x7fbb92b01ae9 in __interceptor_malloc /build/gcc/src/gcc/libsanitizer/asan/asan_malloc_linux.cc:62
    #1 0x7fbb912287ee  (/usr/lib/libfontconfig.so.1+0x1d7ee)

Indirect leak of 96 byte(s) in 3 object(s) allocated from:
    #0 0x7fbb92b01ce1 in __interceptor_calloc /build/gcc/src/gcc/libsanitizer/asan/asan_malloc_linux.cc:70
    #1 0x7fbb91228ed9  (/usr/lib/libfontconfig.so.1+0x1ded9)

Indirect leak of 11 byte(s) in 1 object(s) allocated from:
    #0 0x7fbb92a983b1 in __interceptor_strdup /build/gcc/src/gcc/libsanitizer/asan/asan_interceptors.cc:560
    #1 0x7fbb912281f5 in FcValueSave (/usr/lib/libfontconfig.so.1+0x1d1f5)

```

Reply:
Hm, okay, that's not the most useful backtrace in the world. Still, I would have expected way more leak reports. When I tried to get awesome valgrind-clean, I eventually resorted to suppressions and then gave up.
For example, `valgrind pango-view -t "This is a test"` finds 2 definitely lost blcoks, 304 indirectly lost and 21 possibly lost. The biggest leak has this backtrace:
```
==8098== 24,528 bytes in 3 blocks are still reachable in loss record 948 of 948
==8098==    at 0x4C2FB6B: malloc (vg_replace_malloc.c:299)
==8098==    by 0x5C29924: ??? (in /usr/lib64/libfontconfig.so.1.10.1)
==8098==    by 0x5C43FCD: ??? (in /usr/lib64/libfontconfig.so.1.10.1)
==8098==    by 0x7E14037: ??? (in /usr/lib64/libexpat.so.1.6.7)
==8098==    by 0x7E14E0B: ??? (in /usr/lib64/libexpat.so.1.6.7)
==8098==    by 0x7E12A52: ??? (in /usr/lib64/libexpat.so.1.6.7)
==8098==    by 0x7E13714: ??? (in /usr/lib64/libexpat.so.1.6.7)
==8098==    by 0x7E172DC: XML_ParseBuffer (in /usr/lib64/libexpat.so.1.6.7)
==8098==    by 0x5C42B52: ??? (in /usr/lib64/libfontconfig.so.1.10.1)
==8098==    by 0x5C42F85: FcConfigParseAndLoad (in /usr/lib64/libfontconfig.so.1.10.1)
==8098==    by 0x5C42FF7: FcConfigParseAndLoad (in /usr/lib64/libfontconfig.so.1.10.1)
==8098==    by 0x5C4316D: ??? (in /usr/lib64/libfontconfig.so.1.10.1)
==8098==    by 0x7E14037: ??? (in /usr/lib64/libexpat.so.1.6.7)
==8098==    by 0x7E14E0B: ??? (in /usr/lib64/libexpat.so.1.6.7)
==8098==    by 0x7E12A52: ??? (in /usr/lib64/libexpat.so.1.6.7)
==8098==    by 0x7E13714: ??? (in /usr/lib64/libexpat.so.1.6.7)
==8098==    by 0x7E172DC: XML_ParseBuffer (in /usr/lib64/libexpat.so.1.6.7)
==8098==    by 0x5C42B52: ??? (in /usr/lib64/libfontconfig.so.1.10.1)
==8098==    by 0x5C42F85: FcConfigParseAndLoad (in /usr/lib64/libfontconfig.so.1.10.1)
==8098==    by 0x5C35D53: ??? (in /usr/lib64/libfontconfig.so.1.10.1)
==8098==    by 0x5C35FA5: ??? (in /usr/lib64/libfontconfig.so.1.10.1)
==8098==    by 0x5C29A56: ??? (in /usr/lib64/libfontconfig.so.1.10.1)
==8098==    by 0x5C2BAE4: FcConfigSubstituteWithPat (in /usr/lib64/libfontconfig.so.1.10.1)
==8098==    by 0x54B3C27: ??? (in /usr/lib64/libpangocairo-1.0.so.0.4000.13)
==8098==    by 0x5094761: ??? (in /usr/lib64/libpangoft2-1.0.so.0.4000.13)
==8098==    by 0x4E54416: ??? (in /usr/lib64/libpango-1.0.so.0.4000.13)
==8098==    by 0x4E55B9F: pango_itemize_with_base_dir (in /usr/lib64/libpango-1.0.so.0.4000.13)
==8098==    by 0x4E5E55A: ??? (in /usr/lib64/libpango-1.0.so.0.4000.13)
==8098==    by 0x4E6042B: ??? (in /usr/lib64/libpango-1.0.so.0.4000.13)
==8098==    by 0x4E60900: pango_layout_get_pixel_extents (in /usr/lib64/libpango-1.0.so.0.4000.13)
==8098==    by 0x10D609: ??? (in /usr/bin/pango-view)
==8098==    by 0x10E1C7: ??? (in /usr/bin/pango-view)
==8098==    by 0x110965: ??? (in /usr/bin/pango-view)
==8098==    by 0x10CDDE: ??? (in /usr/bin/pango-view)
==8098==    by 0x66F7009: (below main) (in /usr/lib64/libc-2.26.so)
```
The function `FcConfigParseAndLoad` sounds a lot like it is loading the fontconfig configuration in `/etc/fonts` (this is just a guess!). So I wonder why you aren't seeing this memory leak nor do you need to call `FcFini`.

Also, yes you can use `cairo_debug_reset_static_data` with your own caching, you just have to make sure to clear your own cache before. Actually, your patch does *introduce* a memory leak since your cache is never cleared. ;-)

Finally, feel free to ignore me. @stapelberg asked for some comments and I left them, but I am not the one to decide what goes into i3lock and what not.

Reply:
I’m definitely on board with the approach to first try hard and fix the memory leaks in cairo :)

Reply:
> Also, yes you can use cairo_debug_reset_static_data with your own caching, you just have to make sure to clear your own cache before

ah, yeah, but if we've got to clear our cache every tick, it's not much of a cache, then :P

> Actually, your patch does introduce a memory leak since your cache is never cleared. ;-)

Yeah, I know. There's no real elegant way of doing this all without either caching nothing (a bit less than ideal) or adding some at-exit cleanup to i3lock (probably way more complicated than needed)

Reply:
We’d prefer to not have exit cleanups unless absolutely necessary (e.g. to aid leak-checker tools). The OS cleans up when a process exits, that’s good enough.

Reply:
> The OS cleans up when a process exits, that’s good enough.

Which means that a memory leak is only a bug if it increases over time, not if it is a static cache like it is (or at least seems to be) here?

Reply:
I’d say so, yes. (Again: with the caveat that if a static leak is not recognized as such by a memory leak checker, it might still be good to eliminate it to get a more useful report.)

Reply:
Yeah, I was pretty much making this pull with the goal of having a clean baseline going forwards.